src/main/java/org/sagebionetworks/bridge/exporter3/Exporter3WorkerProcessor.java

```diff
@@ -528,6 +600,127 @@ private String getRawS3KeyForUpload(String appId, String studyId, Upload upload,
         return builder.toString();
     }
 
+    // Creates the upload table row for the upload.
+    // Package-scoped for unit tests.
+    UploadTableRow getUploadTableRow(Upload upload, HealthDataRecordEx3 record, StudyParticipant participant,
+            String recordId, Map<String, String> metadataMap) throws IOException, PollSqsWorkerBadRequestException {
+        UploadTableRow tableRow = null;
+        String assessmentGuid = metadataMap.get(METADATA_KEY_ASSESSMENT_GUID);
+        String appId = record.getAppId();
+
+        if (assessmentGuid != null) {
+            // Top-level parameters.
+            tableRow = new UploadTableRow();
+            tableRow.setAssessmentGuid(assessmentGuid);
+            tableRow.setCreatedOn(upload.getRequestedOn());
+            tableRow.setRecordId(recordId);
+            tableRow.setHealthCode(record.getHealthCode());
+            tableRow.setParticipantVersion(record.getParticipantVersion());
+
+            // Metadata.
+            Map<String, String> tableRowMetadata = new HashMap<>();
+            tableRowMetadata.put(METADATA_KEY_CLIENT_INFO, record.getClientInfo());
```

Comment:
```suggestion
            // TODO: syoung 11/22/2023 Get the client info and user agent from the upload metadata rather than the participant record.
            tableRowMetadata.put(METADATA_KEY_CLIENT_INFO, record.getClientInfo());
```

Ideally, this would be set from the OS version, device, etc from when it was uploaded rather than the last call from the device to the server.

Reply:
We do both. userAgent and clientInfo comes from the User-Agent header when the upload is first created, so it won't get overwritten during redrives. It also doesn't depend on the file format, so it's available for every assessment.

deviceInfo comes from metadata.json and is only available in assessments that we recognize.